civis/pubnub.py

```diff
@@ -0,0 +1,80 @@
+from civis.base import CivisJobFailure, CivisAsyncResultBase, FAILED, DONE
+
+try:
+    from pubnub.pubnub import PubNub
+    from pubnub.pnconfiguration import PNConfiguration
+    from pubnub.callbacks import SubscribeCallback
+    has_pubnub = True
+except ImportError:
+    has_pubnub = False
+
+
+if has_pubnub:
+    class JobCompleteListener(SubscribeCallback):
+        def __init__(self, job_id, callback_function):
+            self.job_id = job_id
+            self.callback_function = callback_function
+
+        def message(self, pubnub, message):
+            try:
+                result = message.message
+                if result['object']['id'] == self.job_id \
+                        and result['run']['state'] in DONE:
+                    self.callback_function()
+            except KeyError:
+                pass
+
+        def status(self, pubnub, status):
+            pass
+
+        def presence(self, pubnub, presence):
+            pass
+
+
+class SubscribableResult(CivisAsyncResultBase):
+    def __init__(self, poller, poller_args, client):
```

Comment:
Other places in the code that need a client object have the argument `api_key=None` and create a new client object from that. It's probably better to follow the same convention here. I think that at least the `client` input should be optional. It's cheap to create a new one.